llvm/lib/Transforms/Scalar/LoopIdiomRecognize.cpp

```diff
@@ -1545,12 +1545,15 @@ static Value *matchShiftULTCondition(BranchInst *BI, BasicBlock *LoopEntry,
   BasicBlock *FalseSucc = BI->getSuccessor(1);
   ICmpInst::Predicate Pred = Cond->getPredicate();
 
-  if (Pred == ICmpInst::ICMP_ULT && FalseSucc == LoopEntry) {
-    Threshold = CmpConst->getZExtValue();
-    return Cond->getOperand(0);
-  }
+  if (Pred != ICmpInst::ICMP_ULT || FalseSucc != LoopEntry)
+    return nullptr;
 
-  return nullptr;
+  std::optional<uint64_t> ValIntOpt = CmpConst->getValue().tryZExtValue();
```

Comment:
So I think this works because in practice the caller only does something useful if the threshold is 2 or 4. However, in future we may want to extend support for recognising idioms with greater thresholds. I do wonder if perhaps it's simpler (and more future-proof) to just return the threshold as a APInt instead of uint64_t? That way you hardly need to change this function at all and just do:

```
  if (Pred == ICmpInst::ICMP_ULT && FalseSucc == LoopEntry) {
    Threshold = CmpConst->getValue();
    return Cond->getOperand(0);
  }
```

The checks for threshold == 2 or threshold == 4 later on should just work when it's a APInt I think. What do you think?

Reply:
Thanks, I agree that makes far more sense - and is simpler! I've updated to change uint64_t -> APInt where necessary.

I've kept the new test, as it seems somewhat useful to keep a test to verify that this update actually fixes the crash that led to the original patch being reverted. Let me know if you disagree? Thanks :)